### PR TITLE
start --parse and --render

### PR DESCRIPTION
**Pull request by lasconic**
- [ ] Fixes #
- [ ] Tests added/updated.
- [ ] `./check.sh` run and eventual issues fixed.
- [ ] Updated CONTRIBUTORS.md, if not already done in the past.

### CODE REVIEW

**Comment by lasconic** [comment]
@BoboTiG How does that look ? Is the approach ok ? Should I move the tests too ?

**Comment by BoboTiG** [comment]
It would be better to not copy all that lang datas. I was thinking of simply import from "script" instead of duplicating everything. And when the new code will be ready, we could do the move and delete completely "scripts". It allows us to continue working on langs without being impacted by the code refactoring.

File: wikidict/parse.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    Dict,
+    Generator,
+    List,
+    Tuple,
+)
+
+if TYPE_CHECKING:  # pragma: nocover
+    from xml.etree.ElementTree import Element
+
+import wikitextparser as wtp
```
Comment (BoboTiG) [inline wikidict/parse.py:18]: All `wikitextparser` stuff can be removed from here, right?

File: wikidict/parse.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    print(f">>> Processing {file} ...", flush=True)
+    for element in xml_iter_parse(str(file)):
+        word, code = xml_parse_element(element)
+        if not word or not code or ":" in word:
+            continue
+
+        if any(head_section in code for head_section in head_sections[locale]):
+            words[word] = code
+        else:
+            print(f"Ignored [{word}]")
```
Comment (BoboTiG) [inline wikidict/parse.py:96]:
```suggestion
            print(f"Ignored [{word}]", flush=True)
```

File: wikidict/parse.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    return words
+
+
+def save(snapshot: str, words: Dict[str, str], output_dir: Path) -> None:
+    """Persist data."""
+    # This file is needed by convert.py
+    raw_data = output_dir / f"data_wikicode-{snapshot}.json"
```
Comment (BoboTiG) [inline wikidict/parse.py:104]: (I do not have a strong opinion on the file name)
```suggestion
    raw_data = output_dir / f"data-wikicode-{snapshot}.json"
```

File: wikidict/render.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        print(">>> No data-XXXXXX.json found. Run with --parse first ... ", flush=True)
+        return 1
+
+    current_file = Path(output_dir, filename)
+
+    print(f">>> Loading {current_file} ...")
+    in_words: Dict[str, str] = load(current_file)
+
+    debug = "DEBUG" in os.environ
+    if debug:
```
Comment (BoboTiG) [inline wikidict/render.py:373]:
```suggestion
    if "DEBUG" in os.environ:
```

File: wikidict/render.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            else:
+                f.write(f"    - {entries[0]!r}\n")
+
+    if templates:
+        with open("templates.txt", "w") as f:
+            for template, entry in sorted(templates.items()):
+                f.write(f"{entry!r} => {template!r}\n")
+
+
+def render(in_words: Dict[str, str], locale: str) -> Words:
```
Comment (BoboTiG) [inline wikidict/render.py:334]: This is here we should try to parallelize stuff. For now, let's do keep it like that, the improvement can be done later.

File: wikidict/render.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+
+def get_latest_json_file(locale: str, output_dir: Path) -> str:
+    """Get the name of the last data-*.json file."""
+    files = os.listdir(output_dir)
+    files = [f for f in files if f.startswith("data_wikicode-") and f.endswith(".json")]
+    if not files:
+        return ""
+    files = sorted(files)
+    return files[-1]
```
Comment (BoboTiG) [inline wikidict/render.py:293]:
```suggestion
    files = list(output_dir.glob("data_wikicode-*.json")
    return sorted(files)[-1] if files else ""
```

**Review by BoboTiG — COMMENTED** [review]
A first quick review. Good job 👍

**Comment by lasconic** [comment]
My python knowledge is limited... How do you solve 
Attempted relative import beyond top-level packagepylint(relative-beyond-top-level)

```
from ..scripts.lang import (
```

File: wikidict/parse.py (hunk 2 above)
Comment (lasconic) [reply wikidict/parse.py:96]: that's a debug statement, I'll remove it

File: wikidict/parse.py (hunk 3 above)
Comment (lasconic) [reply wikidict/parse.py:104]: I like data_wikicode-{snapshot}.json, it separates the name from the date.

**Comment by BoboTiG** [comment]
Ah. If you do not use relative import, does it work (`from scripts.lang import ...`)? This is not important for now, we will come back to relative imports at the end.

File: wikidict/render.py (hunk 4 above)
Comment (lasconic) [reply wikidict/render.py:373]: I will fix this, but I would suggest to create a special --option for this. --find-sections-templates ?

Comment (BoboTiG) [reply wikidict/render.py:373]: Or simply `--find-templates`? But yes, a specific argument would be better.

**Comment by lasconic** [comment]
> Ah. If you do not use relative import, does it work (`from scripts.lang import ...`)? This is not important for now, we will come back to relative imports at the end.

It does...

File: wikidict/render.py (hunk 4 above)
Comment (lasconic) [reply wikidict/render.py]: Ok done. find-templates is now an option, so is --get-word.
wikidict is now on par with scripts.get

File: wikidict/render.py (hunk 6 above)
Comment (BoboTiG) [reply wikidict/render.py:283]: Do you plan to apply those changes?

Comment (lasconic) [reply wikidict/render.py:283]: Sure, checking pool.map now but I will do it next.

Comment (BoboTiG) [reply wikidict/render.py:283]: Oh you are quick :)
That could be done in another PR if you prefer. LMK when you are ready for a final review.

Comment (lasconic) [reply wikidict/render.py:283]: Not so quick :) I couldn't make it work with pool.map yet :(

Comment (lasconic) [reply wikidict/render.py]: Done and squashed.

File: wikidict/render.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+import wikitextparser._spans
+
+# As stated in wikitextparser._spans.parse_pm_pf_tl():
+#   If the byte_array passed to parse_to_spans contains n WikiLinks, then
+#   this function will be called n + 1 times. One time for the whole byte_array
+#   and n times for each of the n WikiLinks.
+#
+# We do not care about links, let's speed-up the all process by skipping the n times call.
+# Doing that is a ~30% optimization.
+wikitextparser._spans.WIKILINK_PARAM_FINDITER = lambda *_: ()
```
Comment (lasconic) [inline wikidict/render.py:35]: Note: I changed this line compared to `get.py`
WIKILINK_FINDITER doesn't exist anymore in wtp.

**Review by BoboTiG — APPROVED** [review]

**Comment by BoboTiG** [comment]
I merge now.

**Pull request merged**